File: 2021/06/main.js
function simulateNaive(fishes, days) {
  if (days === 0) return fishes.length

  return simulateNaive(
    fishes.flatMap((fish) => (fish === 0 ? [6, 8] : fish - 1)),
    days - 1
  )
}

function simulate(fishes, days) {
  const state = new Array(9).fill(0)

  for (const fish of fishes) state[fish]++

  for (let day = 0; day < days; day++) {
    let zeroes = state[0]
    state[0] = state[1]
    state[1] = state[2]
    state[2] = state[3]
    state[3] = state[4]
    state[4] = state[5]
    state[5] = state[6]
    state[6] = state[7] + zeroes
    state[7] = state[8]
    state[8] = zeroes
  }

  return state.reduce((a, b) => a + b, 0)
}

console.log(simulate([3, 4, 3, 1, 2], 80))
console.log(simulate([3, 4, 3, 1, 2], 256))

console.log(simulateNaive([3, 4, 3, 1, 2], 80))
console.log(simulateNaive([3, 4, 3, 1, 2], 256))

File: 2021/06/main.ml
let input = read_line () |> String.split_on_char ',' |> List.map int_of_string

let simulate days fishes =
  let fish = Array.make 9 0 in
  fishes |> List.iter (fun f -> fish.(f) <- fish.(f) + 1);
  for _ = 1 to days do
    let zeroes = fish.(0) in
    fish.(0) <- fish.(1);
    fish.(1) <- fish.(2);
    fish.(2) <- fish.(3);
    fish.(3) <- fish.(4);
    fish.(4) <- fish.(5);
    fish.(5) <- fish.(6);
    fish.(6) <- fish.(7) + zeroes;
    fish.(7) <- fish.(8);
    fish.(8) <- zeroes
  done;
  fish |> Array.fold_left ( + ) 0

let tick = function
  | 0 -> [ 6; 8 ]
  | n -> [ n - 1 ]

let rec simulate_naive days fishes =
  if days = 0
  then List.length fishes
  else simulate_naive (days - 1) (List.concat_map tick fishes)

let () =
  input |> simulate 80 |> Printf.printf "Part 1: %d\n";
  input |> simulate 256 |> Printf.printf "Part 2: %d\n";
  input |> simulate_naive 80 |> Printf.printf "Part 1: %d\n%!";
  input |> simulate_naive 256 |> Printf.printf "Part 2: %d\n"
